### PR TITLE
Validate propTypes for lazy and memo

### PR DESCRIPTION
**Pull request by gaearon**
Fixes https://github.com/facebook/react/issues/14159

### CODE REVIEW

File: packages/react/src/ReactElementValidator.js [hunk 1]
```diff
@@ -177,6 +180,32 @@ function validateChildKeys(node, parentType) {
...
+      case REACT_MEMO_TYPE:
+        innerType = type.type;
+        break;
+      case REACT_LAZY_TYPE: {
+        innerType = refineResolvedLazyComponent(type);
+        break;
+      }
+    }
+    const innerPropTypes = getPropTypes(innerType);
+    return {...innerPropTypes, ...type.propTypes};
```
Comment (sophiebits) [inline packages/react/src/ReactElementValidator.js]: Shouldn’t this intersect if there are two checkers for the same key?

Comment (gaearon) [reply packages/react/src/ReactElementValidator.js]: Hmm. Right. I wasn't sure what's the correct mental model but intersecting makes sense.

Comment (gaearon) [reply packages/react/src/ReactElementValidator.js]: Meh. This seems too complicated. Either would have to always re-create them, or cache them somewhere. I think shadowing is reasonable, no?

**Comment by sophiebits** [comment]
@sebmarkbage Can you explain your thinking here?

**Comment by sebmarkbage** [comment]
You shouldn't be assigning type definitions on the consuming side. This is very different from `memo` which is used on the definition side.

You shouldn't be declaring the definitions of a type at the consuming side since it is not the owner of the interface. In fact, before you load it, you don't necessarily even know what it'll be if it's dynamic.

The argument about parity when you wrap memo or not on the definition side is strong since it's the owner of the code, but the same doesn't apply with `lazy` since `lazy` is on the consuming side. If you just remove lazy, you'd be mutating the definition side.

Further more, we might not even have an special React object here in the future so doesn't seem very future proof. It might e.g. be an AssetReference. At one point is was a Promise. Not sure what the type of `eager` will be once we add that so it'd need to have parity with that.

In fact, we should also warn about `defaultProps` being assigned to the lazy wrapper.

**Comment by sophiebits** [comment]
Makes sense. Thanks.

File: packages/react/src/ReactElementValidator.js (hunk 1 above)
Comment (sebmarkbage) [reply packages/react/src/ReactElementValidator.js]: If shadowing is reasonable, why not the whole set? It should probably be a warning that the type sets are ambiguous and only one should be provided. Seems like an error.

**Comment by gaearon** [comment]
Superseded by https://github.com/facebook/react/pull/14298

**Pull request closed without merging**